### PR TITLE
Add helper functions to QRResult

### PR DESCRIPTION
**Pull request by jmh530**
Addresses issue #17.

Adding functions to produce Q and R matrices as well as reconstruct the original matrix.

Note that this is only tested on square matrices. I could only get qrDecomp function to work on square matrices, which prevented me from testing further.

### CODE REVIEW

File: source/lubeck.d [hunk 1]
```diff
@@ -1960,6 +1960,94 @@ struct QRResult(T)
...
+    auto val_tau = val.tau.slice;
+    
+	auto r = val.R;
+    assert(equal!approxEqual(val.R, R_test));
+    
+    auto q = val.Q;
+    assert(equal!approxEqual(val.Q, Q_test));
+
+    //double-checking values do not change
+	assert(equal!approxEqual(val_matrix, val.matrix));
```
Comment (9il) [inline source/lubeck.d]: tabulation

File: source/lubeck.d [hunk 2]
```diff
@@ -1960,6 +1960,94 @@ struct QRResult(T)
     {
         return qrSolve!(allowDestroy)(matrix, tau, b);
     }
+    
+    /++
+    Extract the Q matrix
+    +/
+    auto Q(Flag!"allowDestroy" allowDestroy = No.allowDestroy)
+    {
+        auto work = [T.sizeof * matrix.length].uninitSlice!T;
```
Comment (9il) [inline source/lubeck.d]: does `T.sizeof` is required here?

File: source/lubeck.d [hunk 3]
```diff
@@ -1960,6 +1960,94 @@ struct QRResult(T)
...
+        
+        auto a = matrix.slice;
```
Comment (9il) [inline source/lubeck.d]: this makes a full copy

File: source/lubeck.d [hunk 4]
```diff
@@ -1960,6 +1960,94 @@ struct QRResult(T)
...
+        auto m = (allowDestroy && a._stride!1 == 1) ? a.assumeCanonical : a.as!T.slice.canonical;
```
Comment (9il) [inline source/lubeck.d]: `a.as!T.slice` makes a copy if the copy.
Well, for stuff like Q, R we can just make a single copy and don't care a lot.

Comment (jmh530) [reply source/lubeck.d]: I just need one copy. I'm removed the extra.

File: source/lubeck.d (hunk 2 above)
Comment (jmh530) [reply source/lubeck.d]: I'll be honest, I really don't know. I had just copied the work variable from qrDecomp. I removed it and it still compiles...

File: source/lubeck.d [hunk 5]
```diff
@@ -1960,6 +1960,93 @@ struct QRResult(T)
...
+            ungqr!T(m, tau, work);
+        return m.transposed;
+    }
+    
+    /++
+    Extract the R matrix
+    +/
+    auto R()
+    {
+        auto r = uninitSlice!T(matrix.shape).universal;
```
Comment (9il) [inline source/lubeck.d]:
```
import mir.algorithm.iteration: eachLower;
auto r = matrix.transposed.slice;  
r.eachLower!"a = 0";
return r; // or r.universal depending on API you like to have
```

File: source/lubeck.d (hunk 2 above)
Comment (9il) [reply source/lubeck.d]: Looks like both are ok. The ideal way is to make a `lwork` query, some mir-lapack functions has workspace size quiery versions like `sysv_rook_wk`, but we didn't yet implemented it for geqrf.

The current version with fixed n can be merged as well.

http://scc.ustc.edu.cn/zlsc/sugon/intel/mkl/mkl_manual/GUID-8E285C9D-50B5-4ADA-8ADA-35FA32FF201A.htm

lda
INTEGER. The leading dimension of a; at least max(1, m).

lwork
INTEGER. The size of the work array (lwork ≥ n).

If lwork = -1, then a workspace query is assumed; the routine only calculates the optimal size of the work array, returns this value as the first entry of the work array, and no error message related to lwork is issued by xerbla.

See Application Notes for the suggested value of lwork.

File: source/lubeck.d (hunk 5 above)
Comment (9il) [reply source/lubeck.d]: hope this will help to fix travis

Comment (jmh530) [reply source/lubeck.d]: This PR is also behind one commit from whatever you fixed. That may also be a factor.

Comment (jmh530) [reply source/lubeck.d]: There will need to be some additional work to get this working with rectangular matrices also. I've noticed some issues already trying to compare the results with what I get in R.

**Comment by jmh530** [comment]
@9il Checks are passing. I still need to make this work for rectangular matrices, but I don't know if you want me to do that in a separate PR.

**Comment by 9il** [comment]
This PR is fine for rectangular changes

**Comment by jmh530** [comment]
@9il Finally got around to getting it working for rectangular matrices. I had tried to get UTs that replicate exactly the qr() output from R, but it didn't match up exactly. The Q() and R() functions though produce the same results, so I think that's sufficient for now.

**Pull request merged**